File: app/src/main/java/com/personal/accountantAssistant/utils/DataBaseUtils.java
package com.personal.accountantAssistant.utils;

import android.app.Activity;
import android.content.Context;

import com.personal.accountantAssistant.R;
import com.personal.accountantAssistant.db.DatabaseManager;
import com.personal.accountantAssistant.ui.payments.entities.Payments;
import com.personal.accountantAssistant.ui.payments.enums.PaymentsType;

import io.reactivex.functions.Action;

public class DataBaseUtils {

    public static boolean isNotDefault(final long idOrRecord) {
        return idOrRecord > Constants.DEFAULT_UID;
    }

    private static DatabaseManager newDatabaseManagerFrom(final Activity activity) {
        final Context context = ActivityUtils.parse(activity);
        return new DatabaseManager(context);
    }

    public static void saveDataFrom(final Activity activity,
                                    final Object entity,
                                    final Action onSuccess) {

        long recordSaved = Constants.DEFAULT_UID;
        final Payments payment = ParserUtils.toPayments(entity);
        final Context context = ActivityUtils.parse(activity);
        final DatabaseManager databaseManager = new DatabaseManager(context);

        if (ParserUtils.isNotNullObject(payment)) {
            recordSaved = databaseManager.insertOrUpdatePayment(payment);
        }

        if (isNotDefault(recordSaved)) {
            CalendarsUtils.createCalendarEventFrom(context, payment);
            ActionUtils.runAction(onSuccess);
        } else {
            CalendarsUtils.deleteCalendarEventsFrom(context, payment);
            ToastUtils.showLongText(context, R.string.error_saving_your_data);
        }
    }

    static void deleteDataFrom(final Context context,
                               final Object entity,
                               final Action onSuccess) {

        long recordDeleted = Constants.DEFAULT_UID;
        final Payments payment = ParserUtils.toPayments(entity);
        final DatabaseManager databaseManager = new DatabaseManager(context);

        if (!ParserUtils.isNullObject(payment)) {
            recordDeleted = databaseManager.deletePaymentsRecordFrom(payment);
        }

        if (isNotDefault(recordDeleted)) {
            CalendarsUtils.deleteCalendarEventsFrom(context, payment);
            ToastUtils.showLongText(context, R.string.successfully_deleted_record);
            ActionUtils.runAction(onSuccess);
        } else {
            ToastUtils.showLongText(context, R.string.error_deleting_records);
        }
    }

    public static void deleteRecord(final Context context,
                                    final Object entity,
                                    final Action onSuccess) {
        DialogUtils.confirmationDialog(context,
                R.string.delete_record_title,
                R.string.delete_record_message,
                () -> deleteDataFrom(context, entity, onSuccess));
    }

    public static void saveDataFromWithoutFinish(final Activity activity,
                                                 final Object entity,
                                                 final Action onSuccess) {
        saveDataFrom(activity, entity, onSuccess);
    }

    public static boolean anyActivePaymentsRecordsBy(final Activity activity,
                                                     final PaymentsType paymentsType) {
        return newDatabaseManagerFrom(activity).anyActivePaymentsRecordsBy(paymentsType);
    }

    public static boolean allActivePaymentsRecordsBy(final Activity activity,
                                                     final PaymentsType paymentsType) {
        return newDatabaseManagerFrom(activity).allActivePaymentsRecordsBy(paymentsType);
    }
}

File: app/src/main/java/com/personal/accountantAssistant/utils/DataBaseUtils.kt
package com.personal.accountantAssistant.utils

import android.app.Activity
import android.content.Context
import android.os.Build
import androidx.annotation.RequiresApi
import com.personal.accountantAssistant.R
import com.personal.accountantAssistant.db.DatabaseManager
import com.personal.accountantAssistant.ui.payments.enums.PaymentsType
import com.personal.accountantAssistant.utils.ActionUtils.runAction
import com.personal.accountantAssistant.utils.ActivityUtils.parse
import com.personal.accountantAssistant.utils.CalendarsUtils.createCalendarEventFrom
import com.personal.accountantAssistant.utils.CalendarsUtils.deleteCalendarEventsFrom
import io.reactivex.functions.Action

object DataBaseUtils {
    @JvmStatic
    fun isNotDefaultRecord(idOrRecord: Long?): Boolean {
        return !isDefaultRecord(idOrRecord)
    }

    @JvmStatic
    fun isDefaultRecord(idOrRecord: Long?): Boolean {
        return idOrRecord?.let { it.toInt() == Constants.DEFAULT_UID } ?: false
    }

    @RequiresApi(Build.VERSION_CODES.P)
    private fun newDatabaseManagerFrom(activity: Activity?): DatabaseManager? {
        return parse(activity)?.let { context ->
            DatabaseManager(context)
        }
    }

    @JvmStatic
    @RequiresApi(Build.VERSION_CODES.P)
    fun saveDataFrom(activity: Activity?, entity: Any?, onSuccess: Action?) {
        val payment = ParserUtils.toPayments(entity)
        parse(activity)?.let { context ->
            val databaseManager = DatabaseManager(context)
            val recordSaved: Long = databaseManager.insertOrUpdatePayment(payment)
            if (isNotDefaultRecord(recordSaved)) {
                createCalendarEventFrom(context, payment)
                onSuccess?.let { runAction(it) }
            } else {
                deleteCalendarEventsFrom(context, payment)
                ToastUtils.showLongText(context, R.string.error_saving_your_data)
            }
        }
    }

    @RequiresApi(Build.VERSION_CODES.P)
    private fun deleteDataFrom(context: Context?, entity: Any?, onSuccess: Action?) {
        val payment = ParserUtils.toPayments(entity)
        context?.let { ctx ->
            val databaseManager = DatabaseManager(ctx)
            val recordDeleted = databaseManager.deletePaymentsRecordFrom(payment)
            if (isNotDefaultRecord(recordDeleted)) {
                deleteCalendarEventsFrom(ctx, payment)
                ToastUtils.showLongText(ctx, R.string.successfully_deleted_record)
                onSuccess?.let { runAction(it) }
            } else {
                ToastUtils.showLongText(ctx, R.string.error_deleting_records)
            }
        }
    }

    @JvmStatic
    @RequiresApi(Build.VERSION_CODES.P)
    fun deleteRecord(context: Context, entity: Any?, onSuccess: Action?) {
        DialogUtils.confirmationDialog(context,
                R.string.delete_record_title,
                R.string.delete_record_message,
                ParserUtils.toAction(deleteDataFrom(context, entity, onSuccess))
        )
    }

    @RequiresApi(Build.VERSION_CODES.P)
    fun saveDataFromWithoutFinish(activity: Activity?, entity: Any?, onSuccess: Action?) {
        saveDataFrom(activity, entity, onSuccess)
    }

    @RequiresApi(Build.VERSION_CODES.P)
    @JvmStatic
    fun anyActivePaymentsRecordsBy(activity: Activity?, paymentsType: PaymentsType?): Boolean? {
        return paymentsType?.let { type -> activity?.let { newDatabaseManagerFrom(it) }?.anyActivePaymentsRecordsBy(type) }
    }

    @JvmStatic
    @RequiresApi(Build.VERSION_CODES.P)
    fun allActivePaymentsRecordsBy(activity: Activity?, paymentsType: PaymentsType?): Boolean? {
        return paymentsType?.let { type -> activity?.let { newDatabaseManagerFrom(it) }?.allActivePaymentsRecordsBy(type) }
    }
}